src/Symfony/Component/Serializer/Debug/Normalizer/AbstractTraceableNormalizer.php

```diff
@@ -0,0 +1,117 @@
+use Symfony\Component\Serializer\SerializerInterface;
+
+abstract class AbstractTraceableNormalizer implements SerializerAwareInterface, DenormalizerAwareInterface, NormalizerAwareInterface, CacheableSupportsMethodInterface
+{
+    /**
+     * @var DenormalizerInterface|NormalizerInterface
+     */
+    protected $delegate;
+    private $normalizations = [];
+    private $denormalizations = [];
+
+    /**
+     * AbstractTraceableNormalizer constructor.
+     *
+     * @param DenormalizerInterface|NormalizerInterface $delegate
+     */
+    public function __construct($delegate)
+    {
+        $this->delegate = $delegate;
+    }
+
+    public function denormalize($data, string $type, string $format = null, array $context = [])
+    {
+        $denormalization = new Denormalization($this->delegate, $data, $type, $format, $context);
+        $denormalization->result = $this->delegate->denormalize($data, $type, $format, $context);
+        $this->denormalizations[] = $denormalization;
+
+        return $denormalization->result;
+    }
+
+    public function supportsDenormalization($data, string $type, string $format = null): bool
+    {
+        return $this->delegate->supportsDenormalization($data, $type, $format);
+    }
+
+    public function normalize($object, string $format = null, array $context = [])
+    {
+        $normalization = new Normalization($this->delegate, $object, $format, $context);
+        $normalization->result = $this->delegate->normalize($object, $format, $context);
+        $this->normalizations[] = $normalization;
+
+        return $normalization->result;
+    }
+
+    public function supportsNormalization($data, string $format = null): bool
+    {
+        return $this->delegate->supportsNormalization($data, $format);
+    }
+
+    /*
+     * Since the Serializer, in its constructor, injects itself into its normalizers,
+     * depending on the implementing interfaces, we need to mimic this behaviour here
+     * and pass them to the delegate.
+     *
+     * Unfortunately this is heavily bound to the Serializer implementation. :(
+     * @see \Symfony\Component\Serializer\Serializer:77
+     */
+
+    public function setSerializer(SerializerInterface $serializer): void
+    {
+        if ($this->delegate instanceof SerializerAwareInterface) {
+            $this->delegate->setSerializer($serializer);
+        }
+    }
+
+    public function setDenormalizer(DenormalizerInterface $denormalizer): void
+    {
+        if ($this->delegate instanceof DenormalizerAwareInterface) {
+            $this->delegate->setDenormalizer($denormalizer);
+        }
+    }
+
+    public function setNormalizer(NormalizerInterface $normalizer): void
+    {
+        if ($this->delegate instanceof NormalizerAwareInterface) {
+            $this->delegate->setNormalizer($normalizer);
+        }
+    }
+
+    public function hasCacheableSupportsMethod(): bool
+    {
+        if ($this->delegate instanceof CacheableSupportsMethodInterface) {
```

Comment:
`return $this->delegate instanceof CacheableSupportsMethodInterface && $this->delegate->hasCacheableSupportsMethod();